Daml codegen for java. (#21215)
digital-asset/daml: sdk/docs/manually-written/sdk/component-howtos/application-development/daml-codegen-java.rst

.. _daml-codegen-java:

Daml Codegen for Java
=====================

> This tool is part of the Daml Assistant and is used to generate Java classes for Daml data types.

Reviewer: ```suggestion

Use the Daml Codegen for Java (``daml codegen java``) to generate Java classes representing all Daml data types defined in a Daml Archive (.dar) file.
These classes simplify constructing the types requires by the Java gRPC bindings for the gRPC Ledger API <TODO: add link to it>; for example, ``com.daml.ledger.api.v1.CreateCommand`` and
``com.daml.ledger.api.v1.ExerciseCommand``.

See "How to work with contracts and transactions in Java" <TODO: add link> for details on how to use the generated classes. See the sections below for guidance on setting up and invoking the codegen.
```

Overview
-------------
You can use the Daml codegen to generate Java classes representing Daml contract templates.
These classes incorporate all boilerplate code for constructing corresponding ledger ``com.daml.ledger.api.v1.CreateCommand``,
``com.daml.ledger.api.v1.ExerciseCommand``, ``com.daml.ledger.api.v1.ExerciseByKeyCommand``, and ``com.daml.ledger.api.v1.CreateAndExerciseCommand``.

Download
-------------

The Daml codegen is part of the Daml Assistant. :ref:`Download the Daml Assistant <assistant-manual-managing-releases>`.

> .. TODO(#347) Fix the link to link to Download the Daml Assistant of "Daml Assistant"

Author: This seems like a typo. Could you check again?

Install
-------------

> The Daml codegen is part of the Daml Assistant.

Reviewer: ```suggestion
Install the Daml Codegen for Java by `installing the Daml Assistant <link-to-daml-assitant-install>`_.
```

(I believe the above is the right syntax for links. The one for the Daml assistant should already be available.)

.. TODO(#347) Add link "Download the Daml Assistant" to point to download section of "Daml Assistant"

Configure
-------------

There are two modes to configure the Daml codegen:

- Command line configuration, specifying **all** settings in the command line.

> - Project file configuration, specifying **all** settings in the ``daml.yaml``.

Reviewer: Note this reads more like reference documentation. Passive, factual voice. 

You can leave it though here, as it doesn't hurt too much here.

However, when writing new text then please follow https://diataxis.fr/how-to-guides/#the-language-of-how-to-guides

Author: rewording

      -o, --output-directory <value>
                               Output directory for the generated sources
      -d, --decoderClass <value>
                               Fully Qualified Class Name of the optional Decoder utility
      -V, --verbosity <value>  Verbosity between 0 (only show errors) and 4 (show all messages) -- defaults to 0
      -r, --root <value>       Regular expression for fully-qualified names of templates to generate -- defaults to .*
      --help                   This help text

Project file configuration
--------------------------

The above settings can be configured in the ``codegen`` element of the Daml project file ``daml.yaml``.

Here is an example::

    sdk-version: 2.0.0
    name: quickstart
    source: daml
    init-script: Main:initialize
    parties:
      - Alice
      - Bob
      - USD_Bank
      - EUR_Bank
    version: 0.0.1
    exposed-modules:
      - Main
    dependencies:
      - daml-prim
      - daml-stdlib
    codegen:
      java:
        package-prefix: com.daml.quickstart.iou
        output-directory: java-codegen/src/main/java
        verbosity: 2

Operate

> -------------

Reviewer: Note: title underlines must match in length.

Author: This is not necessary (they should be "[at least as long as the text"](https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#sections))

Author: I will cut them though since they are prettier if they match in length.

> Run the Daml codegen using project file configuration with::

Author: To keep it uniform with the next code block

<img width="822" alt="image" src="https://github.com/user-attachments/assets/158fdb6c-8693-4a4a-a9bb-78e3e1384741" />
